File: go/test/container_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/playstation-store-api2-sdk/go"
	"github.com/voxgig-sdk/playstation-store-api2-sdk/go/core"

	vs "github.com/voxgig-sdk/playstation-store-api2-sdk/go/utility/struct"
)

func TestContainerEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Container(nil)
		if ent == nil {
			t.Fatal("expected non-nil ContainerEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := containerBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "container." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set PLAYSTATIONSTOREAPI__TEST_CONTAINER_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		containerRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.container", setup.data)))
		var containerRef01Data map[string]any
		if len(containerRef01DataRaw) > 0 {
			containerRef01Data = core.ToMapAny(containerRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = containerRef01Data

		// LIST
		containerRef01Ent := client.Container(nil)
		containerRef01Match := map[string]any{
			"age_limit": setup.idmap["age_limit01"],
			"container_id": setup.idmap["container01"],
			"country": setup.idmap["country01"],
			"language": setup.idmap["language01"],
		}

		containerRef01ListResult, err := containerRef01Ent.List(containerRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, containerRef01ListOk := containerRef01ListResult.([]any)
		if !containerRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", containerRef01ListResult)
		}

	})
}

func containerBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "container", "ContainerTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read container test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse container test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"container01", "container02", "container03", "age_limit01", "country01", "language01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("PLAYSTATIONSTOREAPI__TEST_CONTAINER_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"PLAYSTATIONSTOREAPI__TEST_CONTAINER_ENTID": idmap,
		"PLAYSTATIONSTOREAPI__TEST_LIVE":      "FALSE",
		"PLAYSTATIONSTOREAPI__TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["PLAYSTATIONSTOREAPI__TEST_CONTAINER_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["PLAYSTATIONSTOREAPI__TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewPlaystationStoreApi2SDK(core.ToMapAny(mergedOpts))
	}

	live := env["PLAYSTATIONSTOREAPI__TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["PLAYSTATIONSTOREAPI__TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/ContainerEntityTest.php
<?php
declare(strict_types=1);

// Container entity test

require_once __DIR__ . '/../playstationstoreapi2_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class ContainerEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = PlaystationStoreApi2SDK::test(null, null);
        $ent = $testsdk->Container(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = container_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["list"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "container." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set PLAYSTATIONSTOREAPI__TEST_CONTAINER_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $container_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.container")));
        $container_ref01_data = null;
        if (count($container_ref01_data_raw) > 0) {
            $container_ref01_data = Helpers::to_map($container_ref01_data_raw[0][1]);
        }

        // LIST
        $container_ref01_ent = $client->Container(null);
        $container_ref01_match = [
            "age_limit" => $setup["idmap"]["age_limit01"],
            "container_id" => $setup["idmap"]["container01"],
            "country" => $setup["idmap"]["country01"],
            "language" => $setup["idmap"]["language01"],
        ];

        [$container_ref01_list_result, $err] = $container_ref01_ent->list($container_ref01_match, null);
        $this->assertNull($err);
        $this->assertIsArray($container_ref01_list_result);

    }
}

function container_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/container/ContainerTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = PlaystationStoreApi2SDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["container01", "container02", "container03", "age_limit01", "country01", "language01"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("PLAYSTATIONSTOREAPI__TEST_CONTAINER_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "PLAYSTATIONSTOREAPI__TEST_CONTAINER_ENTID" => $idmap,
        "PLAYSTATIONSTOREAPI__TEST_LIVE" => "FALSE",
        "PLAYSTATIONSTOREAPI__TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["PLAYSTATIONSTOREAPI__TEST_CONTAINER_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["PLAYSTATIONSTOREAPI__TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
            ],
            $extra ?? [],
        ]);
        $client = new PlaystationStoreApi2SDK(Helpers::to_map($merged_opts));
    }

    $live = $env["PLAYSTATIONSTOREAPI__TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["PLAYSTATIONSTOREAPI__TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
